### PR TITLE
v0.25: Task api

### PR DESCRIPTION
**Pull request by maryamsulemani97**

### CODE REVIEW

File: errors.yaml
Comment (gmourier) [inline errors.yaml:69]: This error has been replaced by `invalid_api_key` in the last release. No need to use `invalid_token` now 😇

**Review by gmourier — CHANGES_REQUESTED** [review]

File: learn/advanced/asynchronous_operations.md
Comment (gmourier) [inline learn/advanced/asynchronous_operations.md:134]:
```suggestion
- `failed`: the task failed and nothing has been altered into the database
```

File: reference/api/dump.md
Comment (gmourier) [inline reference/api/dump.md:19]: It would be nice to put a link to the exact error so that the reader doesn't have to scroll down and look for the error on the error page.

File: reference/api/documents.md
Comment (gmourier) [inline reference/api/documents.md:230]:
```suggestion
    "type": "clearAll",
```

File: reference/api/displayed_attributes.md
Comment (gmourier) [inline reference/api/displayed_attributes.md:74]:
```suggestion
    "type": "settingsUpdate",
```

File: reference/api/displayed_attributes.md
Comment (gmourier) [inline reference/api/displayed_attributes.md:108]:
```suggestion
    "type": "settingsUpdate",
```

File: reference/api/distinct_attribute.md
Comment (gmourier) [inline reference/api/distinct_attribute.md:70]:
```suggestion
    "type": "settingsUpdate",
```

File: reference/api/distinct_attribute.md
Comment (gmourier) [inline reference/api/distinct_attribute.md:102]:
```suggestion
    "type": "settingsUpdate",
```

File: reference/api/filterable_attributes.md
Comment (gmourier) [inline reference/api/filterable_attributes.md:69]:
```suggestion
    "type": "settingsUpdate",
```

File: reference/api/filterable_attributes.md
Comment (gmourier) [inline reference/api/filterable_attributes.md:103]:
```suggestion
    "type": "settingsUpdate",
```

File: reference/api/ranking_rules.md
Comment (gmourier) [inline reference/api/ranking_rules.md:86]:
```suggestion
    "type": "settingsUpdate",
```

File: reference/api/ranking_rules.md
Comment (gmourier) [inline reference/api/ranking_rules.md:136]:
```suggestion
    "type": "settingsUpdate",
```

File: reference/api/searchable_attributes.md
Comment (gmourier) [inline reference/api/searchable_attributes.md:80]:
```suggestion
    "type": "settingsUpdate",
```

File: reference/api/searchable_attributes.md
Comment (gmourier) [inline reference/api/searchable_attributes.md:114]:
```suggestion
    "type": "settingsUpdate",
```

File: reference/api/settings.md
Comment (gmourier) [inline reference/api/settings.md:145]:
```suggestion
    "type": "settingsUpdate",
```

File: reference/api/settings.md
Comment (gmourier) [inline reference/api/settings.md:190]:
```suggestion
    "type": "settingsUpdate",
```

File: reference/api/sortable_attributes.md
Comment (gmourier) [inline reference/api/sortable_attributes.md:71]:
```suggestion
    "type": "settingsUpdate",
```

File: reference/api/sortable_attributes.md
Comment (gmourier) [inline reference/api/sortable_attributes.md:105]:
```suggestion
    "type": "settingsUpdate",
```

File: reference/api/stop_words.md
Comment (gmourier) [inline reference/api/stop_words.md:72]:
```suggestion
    "type": "settingsUpdate",
```

File: reference/api/stop_words.md
Comment (gmourier) [inline reference/api/stop_words.md:106]:
```suggestion
    "type": "settingsUpdate",
```

File: reference/api/synonyms.md
Comment (gmourier) [inline reference/api/synonyms.md:78]:
```suggestion
    "type": "settingsUpdate",
```

File: reference/api/synonyms.md
Comment (gmourier) [inline reference/api/synonyms.md:112]:
```suggestion
    "type": "settingsUpdate",
```

File: reference/api/tasks.md
Comment (gmourier) [inline reference/api/tasks.md:39]:
```suggestion
            "details": { 
                    "receivedDocuments": 100,
                    "indexedDocuments": 100
            },
```

File: reference/api/tasks.md
Comment (gmourier) [inline reference/api/tasks.md:140]: It might make more sense to have this part above the previous one.

File: reference/api/tasks.md
Comment (gmourier) [inline reference/api/tasks.md:178]:
```suggestion
            "details": { 
                    "receivedDocuments": 100,
                    "indexedDocuments": 100
            },
```

File: reference/features/dumps.md
Comment (gmourier) [inline reference/features/dumps.md:45]:
```suggestion
Unlike [tasks](/learn/advanced/asynchronous_operations.md), dumps have no queue. **MeiliSearch only processes one dump at a time.** If you attempt to create a dump while another dump is still processing, MeiliSearch will throw an [error](/errors). While a dump is processing, the **task queue is paused and no write operations can occur on the database.** This is also true for [snapshots](/reference/features/snapshots.md#snapshots).
```

**Review by gmourier — CHANGES_REQUESTED** [review]
This is a big piece of change here! Great job!

@maryamsulemani97 for the moment, please don't consider my changes to `numberOfDocuments` in the `details` field of a `task`. 

There are several, I have to confirm with the core team that the final state `details` field e.g `indexedDocuments`/`deletedDocuments` is either returned when the `task` have finished or with a `0`/`null` value while it's still enqueued or being processed.

I will confirm this asap. 😇

File: learn/advanced/asynchronous_operations.md
Comment (gmourier) [reply learn/advanced/asynchronous_operations.md:70]: Hi @maryamsulemani97 👋 

By default, a details field that can have a real value when the task is failed or succeeded will have a `null` value for `enqueued` and `processing` status.

As for `indexedDocuments` field in this example since the `documentsAddition` is `enqueued`.

Comment (maryamsulemani97) [reply learn/advanced/asynchronous_operations.md:70]: Thanks for the update!

File: reference/api/documents.md
Comment (gmourier) [inline reference/api/documents.md:268]: If we put this information here, we should also add this line on all previous and next examples for consistency. 😇

**Review by gmourier — CHANGES_REQUESTED** [review]

File: reference/api/documents.md
Comment (maryamsulemani97) [reply reference/api/documents.md:268]: I do have it for all the examples on this page. Do you mean all the responses under API reference?

Comment (gmourier) [reply reference/api/documents.md:268]: I'm talking about the `line 268` which during my review was only present in this example

Comment (maryamsulemani97) [reply reference/api/documents.md:268]: Sorry if my previous comment wasn’t clear 
I meant I have the line:
>This `uid` allows you to [track the current task](/reference/api/tasks.md).

for all the examples on the /reference/api/documents.md page. You can see that on lines: 145, 205, 235, 268, 306.

Do you want me to add it to all the responses that return a `uid`? E.g. after lines 104, 143, 171 on  /reference/api/indexes.md and other pages under API reference?

Comment (gmourier) [reply reference/api/documents.md:268]: Oh yes, I know what you mean! I don't know what's better actually. I think it's a way to make sure that we indicate that the writings operation are asynchronous.

Is it possible from your side to define reusable snippets to display the same block of information in several places in the documentation code and modify it by making a single modification in one single file?

File: reference/api/tasks.md
Comment (dichotommy) [inline reference/api/tasks.md:13]:
```suggestion
List all tasks globally, regardless of index. The `task` objects are contained in the `results` array.
```

File: reference/api/tasks.md
Comment (dichotommy) [inline reference/api/tasks.md:35]: In the future, we should probably try to make the 0th task always an index creation—doesn't make much sense otherwise. Not a big deal though.

File: reference/api/tasks.md
Comment (dichotommy) [inline reference/api/tasks.md:53]: Removed "by default" because I don't believe there's any way to alter that behavior.

```suggestion
Tasks are displayed in descending order by `uid`. This means that **the most recently created `task` objects appear first**.
```

Comment (dichotommy) [reply reference/api/tasks.md:53]: I think in most cases, sorting by `uid` should be the same as sorting by `enqueuedAt`—it's the order in which requests were received. I wonder if there's a case where this is not true? 🤔

@gmourier tagging you because I'm curious; it's not important so please don't feel any pressure to respond 😇

File: reference/api/tasks.md
Comment (dichotommy) [inline reference/api/tasks.md:56]:
```suggestion
## Get task by `uid`
```

**Review by dichotommy — CHANGES_REQUESTED** [review]
You may still need to add a redirect from `learn/advanced/asynchronous_updates.html` to `learn/advanced/asynchronous_operations.html`

I'm a bit bothered by the way the endpoints are named in the tasks API reference. These endpoints return the whole task object, not just the status field. It's probably true that most people are using the endpoints specifically for checking the status though, so maybe I'm off-base on this. Note that if you do accept my suggestions related to this, you will need to update your links.

Also, based on this, I think you could use a refresher on [the difference between routes and endpoints](https://stackoverflow.com/questions/56075017/difference-between-route-and-endpoint).

File: reference/api/tasks.md
Comment (dichotommy) [inline reference/api/tasks.md:99]:
```suggestion
## Get all tasks for an index
```

File: reference/api/tasks.md
Comment (dichotommy) [inline reference/api/tasks.md:117]:
```suggestion
```

Comment (dichotommy) [reply reference/api/tasks.md:117]: Multiple tasks here, only one is enqueued.

File: reference/api/tasks.md
Comment (dichotommy) [inline reference/api/tasks.md:150]:
```suggestion
## Get task by index
```

File: resources/faq.md
Comment (dichotommy) [inline resources/faq.md:54]: We should maybe replace this FAQ answer w/ something

File: resources/faq.md
Comment (dichotommy) [inline resources/faq.md:157]: This FAQ answer is particularly important, since it's a common problem for first-time users of MeiliSearch—whether because their document addition failed, or because it simply hasn't been processed yet. We definitely need some form of this question in the FAQ.

File: reference/api/tasks.md
Comment (dichotommy) [inline reference/api/tasks.md:60]:
```suggestion
Get a single [task](/learn/advanced/asynchronous_operations.md) by `uid`, regardless of [index](/learn/core_concepts/indexes.md).
```

File: reference/api/tasks.md
Comment (dichotommy) [inline reference/api/tasks.md:103]:
```suggestion
List all [tasks](/learn/advanced/asynchronous_operations.md) for a given [index](/learn/core_concepts/indexes.md).
```

File: reference/api/tasks.md
Comment (dichotommy) [inline reference/api/tasks.md:154]:
```suggestion
Get a single [task](/learn/advanced/asynchronous_operations.md) in a given [index](/learn/core_concepts/indexes.md).
```

File: reference/api/documents.md
Comment (maryamsulemani97) [reply reference/api/documents.md:268]: We don’t have any reusable snippets for now but I'll add this

File: resources/faq.md
Comment (maryamsulemani97) [reply resources/faq.md:54]: I wasn’t sure if we needed something here so I removed it? Maybe explain the summarized task response the users get?

File: learn/getting_started/quick_start.md
Comment (maryamsulemani97) [reply learn/getting_started/quick_start.md:83]: I don’t know where else to add it. I added it here because they will see this response so might as well explain what it does. The only other place is asynchronous operations. I can try fitting it there

File: reference/api/indexes.md
Comment (maryamsulemani97) [reply reference/api/indexes.md:105]: This sentence?

>You can use this `uid` to get more details on [the status of the task](/reference/api/tasks.md#get-task-status-by-uid).

Comment (dichotommy) [reply reference/api/indexes.md:104]: That's the one 😄

File: learn/getting_started/quick_start.md
Comment (dichotommy) [inline learn/getting_started/quick_start.md:90]:
```suggestion
Asynchronous actions return a JSON object containing several fields, the most important of which is `uid`. This indicates that the operation has been taken into account and will be processed once it reaches the front of the queue. To view the task's current `status`, use the [get task endpoint](/reference/api/tasks.md).
```

File: reference/api/README.md
Comment (dichotommy) [inline reference/api/README.md:45]:
```suggestion
**202 - Accepted**: The task has been added to the queue (asynchronous)
```

File: resources/faq.md
Comment (dichotommy) [inline resources/faq.md:54]:
```suggestion
## I did a call to an API route and got an object as a response. What does it mean?
```

**Review by dichotommy — CHANGES_REQUESTED** [review]
This is starting to look really polished 🙌🏻 ⭐ 

I believe you still need to add a redirect for asynchronous operations.

**Comment by maryamsulemani97** [comment]
> I believe you still need to add a redirect for asynchronous operations.

Not sure if I understand this. Do you mean update the links from `/learn/advanced/asynchronous_updates.md` to `learn/advanced/asynchronous_operations.md`? I have already done that.

**Comment by guimachiavelli** [comment]
I believe what @dichotommy means here is that you need to update `.vuepress/public/_redirects`. 

In short, this file tells our server that if someone tries to access e.g. `/learn/advanced/asynchronous_updates.html`, they will be redirected to `/learn/advanced/asynchronous_operations.html` instead. This is necessary because otherwise a person tries to access that resource will simply get a 404 with no instructions of where the content they were looking for was moved.

This might happen when:
1. someone has had that tab open on their browser for ages and tries to refresh it
2. someone uses their browser history to access that page
3. someone clicks on a link outside meili pages (e.g. a programmer posts a meilisearch tutorial on their personal blog) leading to the page we removed/renamed/etc

---

@maryamsulemani97: completely unrelated, but can you ping me when this PR is ready? I want to add a link to the guide on a separate PR, but will wait until this one is fully approved.

**Comment by maryamsulemani97** [comment]
> I believe what @dichotommy means here is that you need to update `.vuepress/public/_redirects`.

Thank you! 🙏 I completely forgot this file
Will let you know when this is approved!

**Review by guimachiavelli — CHANGES_REQUESTED** [review]
Just a couple minor changes and one question I'd like to get a developer's opinion.

Great work! 🌟

**Review by guimachiavelli — APPROVED** [review]
LGTM 🧚

**Comment by guimachiavelli** [comment]
bors merge

File: learn/advanced/asynchronous_operations.md
Comment (gmourier) [reply learn/advanced/asynchronous_operations.md:31]: @guimachiavelli, we have control over what comes out of the API directly. What could possibly happen is that a user uses a particular HTTP client that reorders the JSON on its side. In the specification, the defined order of the fields is obviously for the outgoing response from the API.

I don't know how useful it is to specify it in the documentation. 🤓

**Pull request merged**
